src/utils/codec.ts

```diff
@@ -0,0 +1,418 @@
+    bits: 10,
+    name: 'Base1024',
+    table: BASE_1024,
+    tail: '=',
+}
+
+/**
+ * BASE2048 Encoding
+ */
+export const BASE2048: CodecScheme = {
+    bits: 11,
+    name: 'Base2048',
+    table: parseEmojis(BASE_2048),
+    tail: '=',
+}
+
+/**
+ * Assert `CodecSchema`
+ *
+ * - `bits` : should be `number`
+ * - `table`: `length` of table should be `1 << bits`
+ * - `tail` : should be `string`
+ */
+function assertScheme(scheme: CodecScheme) {
+    // Assert bits
+    if (typeof scheme.bits !== 'number') {
+        throw new Error(`the bits field of ${scheme.name} should be type 'number'`)
+    }
+
+    // Assert table
+    //
+    // Emojis' combining characters needs full of them
+    if (1 << scheme.bits > scheme.table.length) {
+        throw new Error(`table \`length\` of ${scheme.name} should larger than ${1 << scheme.bits}`)
+    }
+
+    // Assert tail
+    if (typeof scheme.tail !== 'string') {
+        throw new Error(`the tail field of ${scheme.name} should be type 'string'`)
+    }
+}
+
+/**
+ * Codec
+ *
+ * Encoding/Decoding string with Unicode code points(0 - 0x10FFFF).
+ *
+ * All data will be formated into binary while constructing Codec
+ * class, for example, Let's format the word `Mask`:
+ *
+ * For `Base1024`:
+ *
+ * | Text |   Binary   | Index |
+ * | ---- | ---------- | ----- |
+ * |      | 0100110101 |  309  |
+ * |      | 1000010111 |  535  |
+ * |      | 0011011011 |  219  |
+ *
+ * For `UTF-16`: `[ 01001101, 01100001, 01110011, 01101011 ]`
+ *
+ * | Text |  Binary  | Index |
+ * | ---- | -------- | ----- |
+ * |  M   | 01001101 |  77   |
+ * |  a   | 01100001 |  97   |
+ * |  s   | 01110011 |  115  |
+ * |  k   | 01101011 |  107  |
+ *
+ * For `Base64`: `[ 010011, 010110, 000101, 110011, 011010, 11 ]`
+ *
+ * | Text | Binary | Index |
+ * | ---- | ------ | ----- |
+ * |  T   | 010011 |  19   |
+ * |  W   | 010110 |  22   |
+ * |  F   | 000101 |  5    |
+ * |  z   | 110011 |  51   |
+ * |  a   | 011010 |  26   |
+ * |  s   | 110000 |  48   |
+ * |  =   | 000000 |       |
+ *
+ * For `hex`: `[ 0100, 1101, 0110, 0001, 0111, 0011, 0110, 1011 ]`
+ *
+ * | Text | Binary | Index |
+ * | ---- | ------ | ----- |
+ * |  4   |  0100  |   4   |
+ * |  d   |  1101  |   13  |
+ * |  6   |  0110  |   6   |
+ * |  1   |  0001  |   1   |
+ * |  7   |  0111  |   7   |
+ * |  3   |  0011  |   3   |
+ * |  6   |  0110  |   6   |
+ * |  b   |  1011  |   11  |
+ *
+ * @property {Point[]} points - Code points
+ *
+ * @method  toHex    - Convert code points to hex string
+ * @method  toBase64 - Convert code points to base64 string
+ * @method  toString - Convert code points to utf-16 string
+ */
+export class Codec {
+    /**
+     * Build `Codec` from target scheme string
+     *
+     * @param  {CodecScheme} scheme - target codec scheme
+     * @param  {string}      input  - the base `CodecScheme` string
+     */
+    static from(scheme: CodecScheme, input: string | string[], emoji: boolean = false): Codec {
```

Comment:
This optional `string | string[]` needs to be optimized.

Reply:
```suggestion
    static from(scheme: CodecScheme, input: string | string[], emoji = false): Codec {
```